feat: Direct writer

yirutang opened the pull request:
Fixes #164 ☕️

Implementation of DirectWriter which allows users to write to BigQuery in one call. Deduplication is not supported but it should be enough for most common write scenarios.

stephaniewang526 commented on google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/SchemaCompact.java:
```diff
@@ -0,0 +1,77 @@
+import com.google.protobuf.Descriptors;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
```
Javadoc on what SchemaCompact does?

yirutang replied:
Done.

stephaniewang526 commented on google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/SchemaCompact.java:
```diff
@@ -0,0 +1,77 @@
+   * @param userSchema
+   */
+  public void check(String tableName, Descriptors.Descriptor userSchema) {
+    com.google.cloud.bigquery.Table table = bigquery.getTable(getTableId(tableName));
```
Could we `import com.google.cloud.bigquery.Table;` so that we could do `Table table = bigquery.getTable(getTableId(tableName));` here?

yirutang replied:
There is another Table defined under v1alpha2, that's why I cannot use Table directly.

stephaniewang526 replied:
if you import the bigquery Table that should give you the bigquery Table though, right?

yirutang replied:
You are right, done.

stephaniewang526 commented on google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/SchemaCompact.java:
```diff
@@ -0,0 +1,77 @@
+   * @param tableName
+   * @param userSchema
+   */
+  public void check(String tableName, Descriptors.Descriptor userSchema) {
```
Could we `import com.google.protobuf.Descriptors.Descriptor;` so we could do `Descriptor userSchema` here?

stephaniewang526 commented on google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/SchemaCompact.java:
```diff
@@ -0,0 +1,77 @@
+    com.google.cloud.bigquery.Table table = bigquery.getTable(getTableId(tableName));
+    Schema schema = table.getDefinition().getSchema();
+    // TODO: We only have very limited check here. More checks to be added.
+    if (schema.getFields().size() != userSchema.getFields().size()) {
```
This seems to be a very barebone check to validate if userSchema is compatible w current schema. Could we at least check field type and field mode? The reason is I see that this check is used in WriterCache.

yirutang replied:
Yes, I don't have time to fully implement the check, more checks will be added. As of now, it just prevents very random input, which, without check, would cause a stream to be created (which will take a day to expire). The more check we have, the happier backend would be. I am planning to have my intern to finish this function.

stephaniewang526 commented on google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/SchemaCompact.java:
```diff
@@ -0,0 +1,77 @@
+   * @param tableName
+   * @param userSchema
+   */
+  public void check(String tableName, Descriptors.Descriptor userSchema) {
```
Why is this a `void` instead of `boolean` return type?

yirutang replied:
I am imagining all the failed case will throw with a exception message.

stephaniewang526 commented on google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/WriterCache.java:
```diff
@@ -0,0 +1,245 @@
+package com.google.cloud.bigquery.storage.v1alpha2;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.protobuf.Descriptors;
```
Could we `import com.google.protobuf.Descriptors.Descriptor;` so that we can use `Descriptor` for all the Descriptor objects in this class?

stephaniewang526 commented on google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/WriterCache.java:
```diff
@@ -0,0 +1,245 @@
+  private static Pattern tablePattern = Pattern.compile(tablePatternString);
+
+  private static WriterCache instance;
+  private LRUCache<String, LRUCache<Descriptors.Descriptor, StreamWriter>> writerCache;
```
```suggestion
  private LRUCache<String, LRUCache<Descriptor, StreamWriter>> writerCache;
```

yirutang replied:
Done.

stephaniewang526 commented on google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/DirectWriter.java:
```diff
@@ -0,0 +1,105 @@
+
+    StreamWriter writer = cache.getTableWriter(tableName, protoRows.get(0).getDescriptorForType());
+    ProtoRows.Builder rowsBuilder = ProtoRows.newBuilder();
+    Descriptors.Descriptor descriptor = null;
```
Similar to other comments below, can we `import com.google.protobuf.Descriptors.Descriptor;` so here we can directly use `Descriptor descriptor`?

yirutang replied:
Done.

stephaniewang526 commented on google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1alpha2/DirectWriterTest.java:
```diff
@@ -0,0 +1,230 @@
+import org.mockito.MockitoAnnotations;
+import org.threeten.bp.Instant;
+
+@RunWith(JUnit4.class)
```
Seeing below Exceptions when running DirectWriterTest. PTAL:
``` Exception in thread "Thread-2" com.google.api.gax.rpc.UnavailableException: io.grpc.StatusRuntimeException: UNAVAILABLE: Channel shutdown invoked
	at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:69)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:72)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:60)
	at com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure(GrpcExceptionCallable.java:97)
	at com.google.api.core.ApiFutures$1.onFailure(ApiFutures.java:68)
	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1039)
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30)
	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1165)
	at com.google.common.util.concurrent.AbstractFuture.addListener(AbstractFuture.java:708)
	at com.google.common.util.concurrent.ForwardingListenableFuture.addListener(ForwardingListenableFuture.java:45)
	at com.google.api.core.ApiFutureToListenableFuture.addListener(ApiFutureToListenableFuture.java:52)
	at com.google.common.util.concurrent.Futures.addCallback(Futures.java:1012)
	at com.google.api.core.ApiFutures.addCallback(ApiFutures.java:63)
	at com.google.api.gax.grpc.GrpcExceptionCallable.futureCall(GrpcExceptionCallable.java:67)
	at com.google.api.gax.rpc.UnaryCallable$1.futureCall(UnaryCallable.java:126)
	at com.google.api.gax.tracing.TracedUnaryCallable.futureCall(TracedUnaryCallable.java:75)
	at com.google.api.gax.rpc.UnaryCallable$1.futureCall(UnaryCallable.java:126)
	at com.google.api.gax.rpc.UnaryCallable.futureCall(UnaryCallable.java:87)
	at com.google.api.gax.rpc.UnaryCallable.call(UnaryCallable.java:112)
	at com.google.cloud.bigquery.storage.v1alpha2.BigQueryWriteClient.createWriteStream(BigQueryWriteClient.java:176)
	at com.google.cloud.bigquery.storage.v1alpha2.WriterCache.CreateNewStream(WriterCache.java:76)
	at com.google.cloud.bigquery.storage.v1alpha2.WriterCache.getTableWriter(WriterCache.java:132)
	at com.google.cloud.bigquery.storage.v1alpha2.DirectWriter.append(DirectWriter.java:79)
	at com.google.cloud.bigquery.storage.v1alpha2.DirectWriterTest$1.run(DirectWriterTest.java:220)
	Suppressed: com.google.api.gax.rpc.AsyncTaskException: Asynchronous task failed
		at com.google.api.gax.rpc.ApiExceptions.callAndTranslateApiException(ApiExceptions.java:57)
		... 6 more
Caused by: io.grpc.StatusRuntimeException: UNAVAILABLE: Channel shutdown invoked
	at io.grpc.Status.asRuntimeException(Status.java:533)
	at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:497)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:426)
	at io.grpc.internal.ClientCallImpl.access$500(ClientCallImpl.java:66)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:689)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$900(ClientCallImpl.java:577)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:751)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:740)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Exception in thread "Thread-6" com.google.api.gax.rpc.UnavailableException: io.grpc.StatusRuntimeException: UNAVAILABLE: Channel shutdown invoked
	at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:69)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:72)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:60)
	at com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure(GrpcExceptionCallable.java:97)
	at com.google.api.core.ApiFutures$1.onFailure(ApiFutures.java:68)
	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1039)
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30)
	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1165)
	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:958)
	at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:749)
	at io.grpc.stub.ClientCalls$GrpcFuture.setException(ClientCalls.java:522)
	at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:497)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:426)
	at io.grpc.internal.ClientCallImpl.access$500(ClientCallImpl.java:66)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:689)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$900(ClientCallImpl.java:577)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:751)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:740)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
	Suppressed: com.google.api.gax.rpc.AsyncTaskException: Asynchronous task failed
		at com.google.api.gax.rpc.ApiExceptions.callAndTranslateApiException(ApiExceptions.java:57)
		at com.google.api.gax.rpc.UnaryCallable.call(UnaryCallable.java:112)
		at com.google.cloud.bigquery.storage.v1alpha2.BigQueryWriteClient.createWriteStream(BigQueryWriteClient.java:176)
		at com.google.cloud.bigquery.storage.v1alpha2.WriterCache.CreateNewStream(WriterCache.java:76)
		at com.google.cloud.bigquery.storage.v1alpha2.WriterCache.getTableWriter(WriterCache.java:132)
		at com.google.cloud.bigquery.storage.v1alpha2.DirectWriter.append(DirectWriter.java:79)
		at com.google.cloud.bigquery.storage.v1alpha2.DirectWriterTest$1.run(DirectWriterTest.java:220)
Caused by: io.grpc.StatusRuntimeException: UNAVAILABLE: Channel shutdown invoked
	at io.grpc.Status.asRuntimeException(Status.java:533)
	... 12 more
Exception in thread "Thread-5" com.google.api.gax.rpc.UnavailableException: io.grpc.StatusRuntimeException: UNAVAILABLE: Channel shutdown invoked
	at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:69)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:72)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:60)
	at com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure(GrpcExceptionCallable.java:97)
	at com.google.api.core.ApiFutures$1.onFailure(ApiFutures.java:68)
	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1039)
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30)
	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1165)
	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:958)
	at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:749)
	at io.grpc.stub.ClientCalls$GrpcFuture.setException(ClientCalls.java:522)
	at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:497)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:426)
	at io.grpc.internal.ClientCallImpl.access$500(ClientCallImpl.java:66)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:689)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$900(ClientCallImpl.java:577)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:751)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:740)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
	Suppressed: com.google.api.gax.rpc.AsyncTaskException: Asynchronous task failed
		at com.google.api.gax.rpc.ApiExceptions.callAndTranslateApiException(ApiExceptions.java:57)
		at com.google.api.gax.rpc.UnaryCallable.call(UnaryCallable.java:112)
		at com.google.cloud.bigquery.storage.v1alpha2.BigQueryWriteClient.createWriteStream(BigQueryWriteClient.java:176)
		at com.google.cloud.bigquery.storage.v1alpha2.WriterCache.CreateNewStream(WriterCache.java:76)
		at com.google.cloud.bigquery.storage.v1alpha2.WriterCache.getTableWriter(WriterCache.java:132)
		at com.google.cloud.bigquery.storage.v1alpha2.DirectWriter.append(DirectWriter.java:79)
		at com.google.cloud.bigquery.storage.v1alpha2.DirectWriterTest$1.run(DirectWriterTest.java:220)
Caused by: io.grpc.StatusRuntimeException: UNAVAILABLE: Channel shutdown invoked
	at io.grpc.Status.asRuntimeException(Status.java:533)
	... 12 more
Exception in thread "Thread-4" com.google.api.gax.rpc.UnavailableException: io.grpc.StatusRuntimeException: UNAVAILABLE: Channel shutdown invoked
	at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:69)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:72)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:60)
	at com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure(GrpcExceptionCallable.java:97)
	at com.google.api.core.ApiFutures$1.onFailure(ApiFutures.java:68)
	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1039)
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30)
	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1165)
	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:958)
	at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:749)
	at io.grpc.stub.ClientCalls$GrpcFuture.setException(ClientCalls.java:522)
	at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:497)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:426)
	at io.grpc.internal.ClientCallImpl.access$500(ClientCallImpl.java:66)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:689)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$900(ClientCallImpl.java:577)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:751)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:740)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
	Suppressed: com.google.api.gax.rpc.AsyncTaskException: Asynchronous task failed
		at com.google.api.gax.rpc.ApiExceptions.callAndTranslateApiException(ApiExceptions.java:57)
		at com.google.api.gax.rpc.UnaryCallable.call(UnaryCallable.java:112)
		at com.google.cloud.bigquery.storage.v1alpha2.BigQueryWriteClient.createWriteStream(BigQueryWriteClient.java:176)
		at com.google.cloud.bigquery.storage.v1alpha2.WriterCache.CreateNewStream(WriterCache.java:76)
		at com.google.cloud.bigquery.storage.v1alpha2.WriterCache.getTableWriter(WriterCache.java:132)
		at com.google.cloud.bigquery.storage.v1alpha2.DirectWriter.append(DirectWriter.java:79)
		at com.google.cloud.bigquery.storage.v1alpha2.DirectWriterTest$1.run(DirectWriterTest.java:220)
Caused by: io.grpc.StatusRuntimeException: UNAVAILABLE: Channel shutdown invoked
	at io.grpc.Status.asRuntimeException(Status.java:533)
	... 12 more
Exception in thread "Thread-3" com.google.api.gax.rpc.UnavailableException: io.grpc.StatusRuntimeException: UNAVAILABLE: Channel shutdown invoked
	at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:69)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:72)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:60)
	at com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure(GrpcExceptionCallable.java:97)
	at com.google.api.core.ApiFutures$1.onFailure(ApiFutures.java:68)
	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1039)
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30)
	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1165)
	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:958)
	at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:749)
	at io.grpc.stub.ClientCalls$GrpcFuture.setException(ClientCalls.java:522)
	at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:497)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:426)
	at io.grpc.internal.ClientCallImpl.access$500(ClientCallImpl.java:66)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:689)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$900(ClientCallImpl.java:577)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:751)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:740)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
	Suppressed: com.google.api.gax.rpc.AsyncTaskException: Asynchronous task failed
		at com.google.api.gax.rpc.ApiExceptions.callAndTranslateApiException(ApiExceptions.java:57)
		at com.google.api.gax.rpc.UnaryCallable.call(UnaryCallable.java:112)
		at com.google.cloud.bigquery.storage.v1alpha2.BigQueryWriteClient.createWriteStream(BigQueryWriteClient.java:176)
		at com.google.cloud.bigquery.storage.v1alpha2.WriterCache.CreateNewStream(WriterCache.java:76)
		at com.google.cloud.bigquery.storage.v1alpha2.WriterCache.getTableWriter(WriterCache.java:132)
		at com.google.cloud.bigquery.storage.v1alpha2.DirectWriter.append(DirectWriter.java:79)
		at com.google.cloud.bigquery.storage.v1alpha2.DirectWriterTest$1.run(DirectWriterTest.java:220)
Caused by: io.grpc.StatusRuntimeException: UNAVAILABLE: Channel shutdown invoked
	at io.grpc.Status.asRuntimeException(Status.java:533)
	... 12 more
Apr 14, 2020 7:07:31 PM com.google.cloud.bigquery.storage.v1alpha2.WriterCache CreateNewStream
INFO: Created write stream:projects/p/datasets/d/tables/t/streams/s
Apr 14, 2020 7:07:31 PM com.google.cloud.bigquery.storage.v1alpha2.StreamWriter refreshAppend
INFO: Write Stream projects/p/datasets/d/tables/t/streams/s connection established
Apr 14, 2020 7:07:31 PM com.google.cloud.bigquery.storage.v1alpha2.WriterCache CreateNewStream
INFO: Created write stream:projects/p/datasets/d/tables/t/streams/s2
Apr 14, 2020 7:07:31 PM com.google.cloud.bigquery.storage.v1alpha2.StreamWriter refreshAppend
INFO: Write Stream projects/p/datasets/d/tables/t/streams/s2 connection established```

yirutang replied:
I couldn't repro it on my local env, maybe it is due to the executor to execute the test is different. I manually created an executor to run the concurrent test.

stephaniewang526 commented on google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/WriterCache.java:
```diff
@@ -0,0 +1,245 @@
+    StreamWriter writer = null;
+    LRUCache<Descriptors.Descriptor, StreamWriter> tableEntry = null;
+
+    synchronized (this) {
```
I found this logic a bit hard to read and wonder if we could refactor into this:
``` synchronized (this) {
      tableEntry = writerCache.get(tableName);
      if (tableEntry != null) {
        writer = tableEntry.get(userSchema);
        if (writer != null) {
          if (writer.expired())
            writer.close();
        return writer;
        } else {
          compact.check(tableName, userSchema);
          streamName = CreateNewStream(tableName);
          writer = CreateNewWriter(streamName);
          tableEntry.put(userSchema, writer);
        }
      } else {
        compact.check(tableName, userSchema);
        streamName = CreateNewStream(tableName);
        writer = CreateNewWriter(streamName);
        tableEntry = new LRUCache<Descriptors.Descriptor, StreamWriter>(MAX_WRITERS_PER_TABLE);
        tableEntry.put(userSchema, writer);
        writerCache.put(tableName, tableEntry);
      }
    } ```

yirutang replied:
Done.

stephaniewang526 commented on google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1alpha2/WriterCacheTest.java:
```diff
@@ -0,0 +1,283 @@
+import org.threeten.bp.Instant;
+import org.threeten.bp.temporal.ChronoUnit;
+
+@RunWith(JUnit4.class)
```
Seeing below Exceptions when running WriterCacheTest class. PTAL:
```
Exception in thread "Thread-16" com.google.api.gax.rpc.UnavailableException: io.grpc.StatusRuntimeException: UNAVAILABLE: Channel shutdown invoked
	at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:69)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:72)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:60)
	at com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure(GrpcExceptionCallable.java:97)
	at com.google.api.core.ApiFutures$1.onFailure(ApiFutures.java:68)
	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1039)
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30)
	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1165)
	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:958)
	at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:749)
	at io.grpc.stub.ClientCalls$GrpcFuture.setException(ClientCalls.java:522)
	at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:497)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:426)
	at io.grpc.internal.ClientCallImpl.access$500(ClientCallImpl.java:66)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:689)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$900(ClientCallImpl.java:577)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:751)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:740)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
	Suppressed: com.google.api.gax.rpc.AsyncTaskException: Asynchronous task failed
		at com.google.api.gax.rpc.ApiExceptions.callAndTranslateApiException(ApiExceptions.java:57)

		at com.google.api.gax.rpc.UnaryCallable.call(UnaryCallable.java:112)
		at com.google.cloud.bigquery.storage.v1alpha2.BigQueryWriteClient.createWriteStream(BigQueryWriteClient.java:176)
		at com.google.cloud.bigquery.storage.v1alpha2.WriterCache.CreateNewStream(WriterCache.java:76)
		at com.google.cloud.bigquery.storage.v1alpha2.WriterCache.getTableWriter(WriterCache.java:132)
		at com.google.cloud.bigquery.storage.v1alpha2.WriterCacheTest$1.run(WriterCacheTest.java:274)
Caused by: io.grpc.StatusRuntimeException: UNAVAILABLE: Channel shutdown invoked
	at io.grpc.Status.asRuntimeException(Status.java:533)
	... 12 more
Exception in thread "Thread-24" Exception in thread "Thread-25" com.google.api.gax.rpc.UnavailableException: io.grpc.StatusRuntimeException: UNAVAILABLE: Channel shutdown invoked

```

yirutang replied:
Change it to executors.

stephaniewang526 commented on google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/WriterCache.java:
```diff
@@ -0,0 +1,242 @@
+    }
+  }
+
+  private static class LRUCache<K, V> {
```
Have you considered using [Guava](https://github.com/google/guava/wiki/CachesExplained)? We should not be writing our own LRU cahce.

stephaniewang526 commented on google-cloud-bigquerystorage/.flattened-pom.xml:
```diff
@@ -0,0 +1,1206 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Please don't commit this hidden file.

yirutang replied:
Done

stephaniewang526 commented:
@yirutang Could you look at the two flaky tests that are breaking the build:
https://source.cloud.google.com/results/invocations/eb9b4e28-07c8-4d19-b259-f9ff65354df0/targets/github%2Fjava-bigquerystorage%2Fgoogle-cloud-bigquerystorage%2Ftarget%2Ffailsafe-reports%2Fintegration%2Fcom.google.cloud.bigquery.storage.v1alpha2.it.ITBigQueryWriteManualClientTest/tests;group=com.google.cloud.bigquery.storage.v1alpha2.it.ITBigQueryWriteManualClientTest%28sponge_log%29;test=testComplicateSchemaWithPendingStream;row=1

yirutang commented:
Yes, I fixed the second e2e test, for the first one, it didn't fail on my
machine but maybe there is a timing issue. I added some sleep to the test.

On Wed, Apr 15, 2020 at 2:46 PM Stephanie Wang <notifications@github.com>
wrote:

> @yirutang <https://github.com/yirutang> Could you look at the two flaky
> tests that are breaking the build:
>
> https://source.cloud.google.com/results/invocations/eb9b4e28-07c8-4d19-b259-f9ff65354df0/targets/github%2Fjava-bigquerystorage%2Fgoogle-cloud-bigquerystorage%2Ftarget%2Ffailsafe-reports%2Fintegration%2Fcom.google.cloud.bigquery.storage.v1alpha2.it.ITBigQueryWriteManualClientTest/tests;group=com.google.cloud.bigquery.storage.v1alpha2.it.ITBigQueryWriteManualClientTest%28sponge_log%29;test=testComplicateSchemaWithPendingStream;row=1
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/googleapis/java-bigquerystorage/pull/165#issuecomment-614296498>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHGYVEWTETDOG2FA2TDEKWLRMYTMJANCNFSM4MHF6LJQ>
> .
>


-- 
Thanks.

Yiru

yirutang commented:
Please try rerun the tests, thanks!

On Wed, Apr 15, 2020 at 4:35 PM Yiru Tang <yiru@google.com> wrote:

> Yes, I fixed the second e2e test, for the first one, it didn't fail on my
> machine but maybe there is a timing issue. I added some sleep to the test.
>
> On Wed, Apr 15, 2020 at 2:46 PM Stephanie Wang <notifications@github.com>
> wrote:
>
>> @yirutang <https://github.com/yirutang> Could you look at the two flaky
>> tests that are breaking the build:
>>
>> https://source.cloud.google.com/results/invocations/eb9b4e28-07c8-4d19-b259-f9ff65354df0/targets/github%2Fjava-bigquerystorage%2Fgoogle-cloud-bigquerystorage%2Ftarget%2Ffailsafe-reports%2Fintegration%2Fcom.google.cloud.bigquery.storage.v1alpha2.it.ITBigQueryWriteManualClientTest/tests;group=com.google.cloud.bigquery.storage.v1alpha2.it.ITBigQueryWriteManualClientTest%28sponge_log%29;test=testComplicateSchemaWithPendingStream;row=1
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/googleapis/java-bigquerystorage/pull/165#issuecomment-614296498>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AHGYVEWTETDOG2FA2TDEKWLRMYTMJANCNFSM4MHF6LJQ>
>> .
>>
>
>
> --
> Thanks.
>
> Yiru
>


-- 
Thanks.

Yiru

stephaniewang526 commented:
> Yes, I fixed the second e2e test, for the first one, it didn't fail on my machine but maybe there is a timing issue. I added some sleep to the test.
> […](#)
> On Wed, Apr 15, 2020 at 2:46 PM Stephanie Wang ***@***.***> wrote: @yirutang <https://github.com/yirutang> Could you look at the two flaky tests that are breaking the build: https://source.cloud.google.com/results/invocations/eb9b4e28-07c8-4d19-b259-f9ff65354df0/targets/github%2Fjava-bigquerystorage%2Fgoogle-cloud-bigquerystorage%2Ftarget%2Ffailsafe-reports%2Fintegration%2Fcom.google.cloud.bigquery.storage.v1alpha2.it.ITBigQueryWriteManualClientTest/tests;group=com.google.cloud.bigquery.storage.v1alpha2.it.ITBigQueryWriteManualClientTest%28sponge_log%29;test=testComplicateSchemaWithPendingStream;row=1 — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#165 (comment)](https://github.com/googleapis/java-bigquerystorage/pull/165#issuecomment-614296498)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AHGYVEWTETDOG2FA2TDEKWLRMYTMJANCNFSM4MHF6LJQ> .
> -- Thanks. Yiru

https://source.cloud.google.com/results/invocations/8cba25ad-8919-4fdd-8446-d00222bfb028/targets/github%2Fjava-bigquerystorage%2Fgoogle-cloud-bigquerystorage%2Ftarget%2Ffailsafe-reports%2Fintegration%2Fcom.google.cloud.bigquery.storage.v1beta2.it.ITBigQueryStorageLongRunningTest/tests 

those two test seem to be running better now but this one is flaky.

yirutang commented:
I increased the sleep time a bit longer. Basically, there could be a small
period of time when the data is streamed and when it showed up in reading.

On Wed, Apr 15, 2020 at 4:46 PM Stephanie Wang <notifications@github.com>
wrote:

> Yes, I fixed the second e2e test, for the first one, it didn't fail on my
> machine but maybe there is a timing issue. I added some sleep to the test.
> … <#m_28468367480736070_>
> On Wed, Apr 15, 2020 at 2:46 PM Stephanie Wang *@*.***> wrote: @yirutang
> <https://github.com/yirutang> https://github.com/yirutang Could you look
> at the two flaky tests that are breaking the build:
> https://source.cloud.google.com/results/invocations/eb9b4e28-07c8-4d19-b259-f9ff65354df0/targets/github%2Fjava-bigquerystorage%2Fgoogle-cloud-bigquerystorage%2Ftarget%2Ffailsafe-reports%2Fintegration%2Fcom.google.cloud.bigquery.storage.v1alpha2.it.ITBigQueryWriteManualClientTest/tests;group=com.google.cloud.bigquery.storage.v1alpha2.it.ITBigQueryWriteManualClientTest%28sponge_log%29;test=testComplicateSchemaWithPendingStream;row=1
> — You are receiving this because you were mentioned. Reply to this email
> directly, view it on GitHub <#165 (comment)
> <https://github.com/googleapis/java-bigquerystorage/pull/165#issuecomment-614296498>>,
> or unsubscribe
> https://github.com/notifications/unsubscribe-auth/AHGYVEWTETDOG2FA2TDEKWLRMYTMJANCNFSM4MHF6LJQ
> .
> -- Thanks. Yiru
>
>
> https://source.cloud.google.com/results/invocations/8cba25ad-8919-4fdd-8446-d00222bfb028/targets/github%2Fjava-bigquerystorage%2Fgoogle-cloud-bigquerystorage%2Ftarget%2Ffailsafe-reports%2Fintegration%2Fcom.google.cloud.bigquery.storage.v1beta2.it.ITBigQueryStorageLongRunningTest/tests
>
> those two test seem to be running better now but this one is flaky.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/googleapis/java-bigquerystorage/pull/165#issuecomment-614334316>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHGYVEWVR5FAHD56FUXMR4DRMZBM7ANCNFSM4MHF6LJQ>
> .
>


-- 
Thanks.

Yiru

stephaniewang526 commented:
PTAL:
1. https://source.cloud.google.com/results/invocations/b734458e-0eaf-486f-ae01-c0489f092ca7/targets/github%2Fjava-bigquerystorage%2Fgoogle-cloud-bigquerystorage%2Ftarget%2Ffailsafe-reports%2Fintegration%2Fcom.google.cloud.bigquery.storage.v1alpha2.it.ITBigQueryWriteManualClientTest/tests

2. https://source.cloud.google.com/results/invocations/87754505-3f27-43e1-97ec-4a3ee70b4705/targets (to fix this, run `mvn com.coveo:fmt-maven-plugin:format`)

yirutang commented:
I think I should fix the tests this time. Please verify, thanks.

On Thu, Apr 16, 2020 at 9:06 AM Stephanie Wang <notifications@github.com>
wrote:

> PTAL:
>
>    1.
>
>
>    https://source.cloud.google.com/results/invocations/b734458e-0eaf-486f-ae01-c0489f092ca7/targets/github%2Fjava-bigquerystorage%2Fgoogle-cloud-bigquerystorage%2Ftarget%2Ffailsafe-reports%2Fintegration%2Fcom.google.cloud.bigquery.storage.v1alpha2.it.ITBigQueryWriteManualClientTest/tests
>    2.
>
>
>    https://source.cloud.google.com/results/invocations/87754505-3f27-43e1-97ec-4a3ee70b4705/targets
>    (to fix this, run mvn com.coveo:fmt-maven-plugin:format)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/googleapis/java-bigquerystorage/pull/165#issuecomment-614745483>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHGYVEVFJU56L2YDVHGKDNLRM4UJFANCNFSM4MHF6LJQ>
> .
>


-- 
Thanks.

Yiru
